In [5]:
import opensmile
import pandas as pd

In [6]:
df = pd.read_csv('dataset.csv')
df.head()
example_path = df['file_path'][0]

### Exploring the number of features

In [3]:
smile = opensmile.Smile(
    feature_set=opensmile.FeatureSet.eGeMAPSv02,
    feature_level=opensmile.FeatureLevel.Functionals,
)

y = smile.process_file(example_path)
y.shape

(1, 88)

In [4]:
smile = opensmile.Smile(
    feature_set=opensmile.FeatureSet.GeMAPSv01b,
    feature_level=opensmile.FeatureLevel.Functionals,
)

y = smile.process_file(example_path)
y.shape

(1, 62)

In [5]:
smile = opensmile.Smile(
    feature_set=opensmile.FeatureSet.emobase ,
    feature_level=opensmile.FeatureLevel.Functionals,
)

y = smile.process_file(example_path)
y.shape

(1, 988)

In [6]:
smile = opensmile.Smile(
    feature_set=opensmile.FeatureSet.ComParE_2016 ,
    feature_level=opensmile.FeatureLevel.Functionals,
)

y = smile.process_file(example_path)
y.shape

(1, 6373)

### Process with eGeMAPSv02

In [7]:
smile = opensmile.Smile(
    feature_set=opensmile.FeatureSet.eGeMAPSv02,
    feature_level=opensmile.FeatureLevel.Functionals,
)

y = smile.process_file(example_path)

In [8]:
from tqdm import tqdm
def generate_df(file_paths):
    dfs = []  # List to accumulate DataFrames
    
    for path in tqdm(file_paths, desc="Processing files"):
        # Assuming smile.process_file returns a DataFrame
        y = smile.process_file(path)
        dfs.append(y)  # Append the DataFrame to the list
    
    # Concatenate all DataFrames in the list
    concatenated_df = pd.concat(dfs, ignore_index=True)
    
    return concatenated_df

In [9]:
df_processed = generate_df(df['file_path'])

Processing files: 100%|████████████████████████████████████████████████████████████| 1975/1975 [03:45<00:00,  8.78it/s]


In [10]:
labels = df['emotion'].unique()
labels.sort()
num_clusters = len(labels)

In [11]:
import os

# Set LOKY_MAX_CPU_COUNT environment variable
os.environ["LOKY_MAX_CPU_COUNT"] = "4"  # Replace "4" with the number of cores you want to use

from sklearn.cluster import KMeans

X_train = df_processed.values
kmeans = KMeans(n_clusters=num_clusters, random_state=42)
kmeans.fit(X_train)

# Get cluster labels for each data point
cluster_labels = kmeans.labels_

from sklearn.metrics import adjusted_rand_score

# Assuming y_true contains the true labels for your data points
# cluster_labels contains the cluster assignments obtained from K-means clustering

ari = adjusted_rand_score(df['emotion'], cluster_labels)
print("Adjusted Rand Index (ARI):", ari)

Adjusted Rand Index (ARI): 0.03947090366749895


In [12]:
df['emotion'][:20].values

array(['happy', 'neutral', 'angry', 'happy', 'neutral', 'sad', 'angry',
       'angry', 'fearful', 'happy', 'boredom', 'neutral', 'sad', 'angry',
       'fearful', 'happy', 'neutral', 'sad', 'angry', 'angry'],
      dtype=object)

In [13]:
cluster_labels[:20]

array([2, 1, 2, 2, 1, 1, 2, 2, 2, 2, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2])

This is an extremely low level of agreement on 88 features.

### Try scaling df_preprocess

In [14]:
from sklearn.preprocessing import StandardScaler

# Instantiate the StandardScaler
scaler = StandardScaler()

# Fit the scaler to your data
scaler.fit(df_processed)

# Transform the data using the scaler
df_scaled = scaler.transform(df_processed)

In [15]:
X_train = df_scaled
kmeans = KMeans(n_clusters=num_clusters, random_state=42)
kmeans.fit(X_train)

# Get cluster labels for each data point
cluster_labels = kmeans.labels_

from sklearn.metrics import adjusted_rand_score

# Assuming y_true contains the true labels for your data points
# cluster_labels contains the cluster assignments obtained from K-means clustering

ari = adjusted_rand_score(df['emotion'], cluster_labels)
print("Adjusted Rand Index (ARI):", ari)
print()
print('Actual Labels:',df['emotion'][:20].values)
print()
print('Unsupervised Labels', cluster_labels[:20])

Adjusted Rand Index (ARI): 0.09575353194440772

Actual Labels: ['happy' 'neutral' 'angry' 'happy' 'neutral' 'sad' 'angry' 'angry'
 'fearful' 'happy' 'boredom' 'neutral' 'sad' 'angry' 'fearful' 'happy'
 'neutral' 'sad' 'angry' 'angry']

Unsupervised Labels [4 3 8 8 3 3 8 8 8 8 3 3 2 8 4 4 3 2 8 8]


### Results
1. Scaling improves unsupervised K-Means Clustering from 0.03 to 0.09 however, 0.09 is still a bad score.

### Check processing

In [16]:
df_processed

F0semitoneFrom27.5Hz_sma3nz_amean  \
0                             31.188166   
1                             25.022938   
2                             34.292320   
3                             37.576572   
4                             25.040190   
...                                 ...   
1970                          44.332615   
1971                          43.539474   
1972                          45.137974   
1973                          43.194508   
1974                          44.371799   

      F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
0                                   0.152560   
1                                   0.148540   
2                                   0.102067   
3                                   0.108607   
4                                   0.116655   
...                                      ...   
1970                                0.118625   
1971                                0.138447   
1972                                0.160381   
1973                                0.098176   
1974                                0.112749   

      F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
0                                      27.766502   
1                                      20.637066   
2                                      30.504089   
3                                      34.083504   
4                                      21.225212   
...                                          ...   
1970                                   38.580246   
1971                                   38.441360   
1972                                   38.204502   
1973                                   40.304451   
1974                                   39.802952   

      F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
0                                      31.412098   
1                                      25.118477   
2                                      33.584766   
3                                      37.371193   
4                                      26.198669   
...                                          ...   
1970                                   43.328392   
1971                                   41.650421   
1972                                   43.318497   
1973                                   42.676632   
1974                                   44.422417   

      F0semitoneFrom27.5Hz_sma3nz_percentile80.0  \
0                                      36.340343   
1                                      28.660353   
2                                      37.343616   
3                                      41.233379   
4                                      27.725224   
...                                          ...   
1970                                   49.545036   
1971                                   50.412521   
1972                                   54.491749   
1973                                   45.354191   
1974                                   48.187508   

      F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2  \
0                                     8.573841   
1                                     8.023287   
2                                     6.839527   
3                                     7.149876   
4                                     6.500011   
...                                        ...   
1970                                 10.964790   
1971                                 11.971161   
1972                                 16.287247   
1973                                  5.049740   
1974                                  8.384556   

      F0semitoneFrom27.5Hz_sma3nz_meanRisingSlope  \
0                                       34.608952   
1                                       54.014923   
2                                       35.548016   
3                                       30.146019   
4                                       19.461380   
...                                           ...   
1970                                    29.118887   
1971        

In [17]:
y = smile.process_file(df['file_path'][0])
y

F0semitoneFrom27.5Hz_sma3nz_amean  \
file             start  end                                                         
data/03a01Fa.wav 0 days 0 days 00:00:01.898250                          31.188166   

                                                F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file             start  end                                                              
data/03a01Fa.wav 0 days 0 days 00:00:01.898250                                 0.15256   

                                                F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file             start  end                                                                  
data/03a01Fa.wav 0 days 0 days 00:00:01.898250                                   27.766502   

                                                F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
file             start  end                                                                  
data/03a01Fa.wav 0 days 0 days 00:00:01.898250                                   31.412098   

                                                F0semitoneFrom27.5Hz_sma3nz_percentile80.0  \
file             start  end                                                                  
data/03a01Fa.wav 0 days 0 days 00:00:01.898250                                   36.340343   

                                                F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2  \
file             start  end                                                                
data/03a01Fa.wav 0 days 0 days 00:00:01.898250                                  8.573841   

                                                F0semitoneFrom27.5Hz_sma3nz_meanRisingSlope  \
file             start  end                                                                   
data/03a01Fa.wav 0 days 0 days 00:00:01.898250                                    34.608952   

                                                F0semitoneFrom27.5Hz_sma3nz_stddevRisingSlope  \
file             start  end                                                                     
data/03a01Fa.wav 0 days 0 days 00:00:01.898250                                      11.036856   

                                                F0semitoneFrom27.5Hz_sma3nz_meanFallingSlope  \
file             start  end                                                                    
data/03a01Fa.wav 0 days 0 days 00:00:01.898250                                     91.834801   

                                                F0semitoneFrom27.5Hz_sma3nz_stddevFallingSlope  \
file             start  end                                                                      
data/03a01Fa.wav 0 days 0 days 00:00:01.898250                                       43.389225   

                                                ...  \
file             start  end                     ...   
data/03a01Fa.wav 0 days 0 days 00:00:01.898250  ...   

                                                slopeUV0-500_sma3nz_amean  \
file             start  end                                                 
data/03a01Fa.wav 0 days 0 days 00:00:01.898250                  -0.012351   

                                                slopeUV500-1500_sma3nz_amean  \
file             start  end                                                    
data/03a01Fa.wav 0 days 0 days 00:00:01.898250                      0.010897   

                                                spectralFluxUV_sma3nz_amean  \
file             start  end                                                   
data/03a01Fa.wav 0 days 0 days 00:00:01.898250                     0.218922   

                                                loudnessPeaksPerSec  \
file             start  end                                           
data/03a01Fa.wav 0 days 0 days 00:00:01.898250             4.255319   

                                                VoicedSegmentsPerSec  \
file             start  end                                           

In [18]:
y = smile.process_file(df['file_path'][1])
y

F0semitoneFrom27.5Hz_sma3nz_amean  \
file             start  end                                                         
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                          25.022938   

                                                F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file             start  end                                                              
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                                 0.14854   

                                                F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file             start  end                                                                  
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                                   20.637066   

                                                F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
file             start  end                                                                  
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                                   25.118477   

                                                F0semitoneFrom27.5Hz_sma3nz_percentile80.0  \
file             start  end                                                                  
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                                   28.660353   

                                                F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2  \
file             start  end                                                                
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                                  8.023287   

                                                F0semitoneFrom27.5Hz_sma3nz_meanRisingSlope  \
file             start  end                                                                   
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                                    54.014923   

                                                F0semitoneFrom27.5Hz_sma3nz_stddevRisingSlope  \
file             start  end                                                                     
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                                      38.583805   

                                                F0semitoneFrom27.5Hz_sma3nz_meanFallingSlope  \
file             start  end                                                                    
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                                      27.07468   

                                                F0semitoneFrom27.5Hz_sma3nz_stddevFallingSlope  \
file             start  end                                                                      
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                                        0.593055   

                                                ...  \
file             start  end                     ...   
data/03a01Nc.wav 0 days 0 days 00:00:01.611250  ...   

                                                slopeUV0-500_sma3nz_amean  \
file             start  end                                                 
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                  -0.020058   

                                                slopeUV500-1500_sma3nz_amean  \
file             start  end                                                    
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                      0.007849   

                                                spectralFluxUV_sma3nz_amean  \
file             start  end                                                   
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                     0.450747   

                                                loudnessPeaksPerSec  \
file             start  end                                           
data/03a01Nc.wav 0 days 0 days 00:00:01.611250                 6.25   

                                                VoicedSegmentsPerSec  \
file             start  end                                           

In [19]:
y = smile.process_file(df['file_path'][1973])
y

F0semitoneFrom27.5Hz_sma3nz_amean  \
file                                   start  end                                                            
data/Actor_24_03-01-08-02-02-01-24.wav 0 days 0 days 00:00:03.670333333                          43.194508   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file                                   start  end                                                                 
data/Actor_24_03-01-08-02-02-01-24.wav 0 days 0 days 00:00:03.670333333                                0.098176   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file                                   start  end                                                                     
data/Actor_24_03-01-08-02-02-01-24.wav 0 days 0 days 00:00:03.670333333                                   40.304451   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
file                                   start  end                                                                     
data/Actor_24_03-01-08-02-02-01-24.wav 0 days 0 days 00:00:03.670333333                                   42.676632   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile80.0  \
file                                   start  end                                                                     
data/Actor_24_03-01-08-02-02-01-24.wav 0 days 0 days 00:00:03.670333333                                   45.354191   

                                                                         F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2  \
file                                   start  end                                                                   
data/Actor_24_03-01-08-02-02-01-24.wav 0 days 0 days 00:00:03.670333333                                   5.04974   

                                                                         F0semitoneFrom27.5Hz_sma3nz_meanRisingSlope  \
file                                   start  end                                                                      
data/Actor_24_03-01-08-02-02-01-24.wav 0 days 0 days 00:00:03.670333333                                   155.292587   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevRisingSlope  \
file                                   start  end                                                                        
data/Actor_24_03-01-08-02-02-01-24.wav 0 days 0 days 00:00:03.670333333                                     257.793976   

                                                                         F0semitoneFrom27.5Hz_sma3nz_meanFallingSlope  \
file                                   start  end                                                                       
data/Actor_24_03-01-08-02-02-01-24.wav 0 days 0 days 00:00:03.670333333                                    401.537354   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevFallingSlope  \
file                                   start  end                                                                         
data/Actor_24_03-01-08-02-02-01-24.wav 0 days 0 days 00:00:03.670333333                                      709.714722   

                                                                         ...  \
file                                   start  end                        ...   
data/Actor_24_03-01-08-02-02-01-24.wav 0 days 0 days 00:00:03.670333333  ...   

                                                                         slopeUV0-500_sma3nz_amean  \
file                                   start  end                                                    
data/Actor_24_03-01-08-02-02-01-24.wav 0 days 0 days 00:00:03.6

In [20]:
y = smile.process_file(df['file_path'][1974])
y

F0semitoneFrom27.5Hz_sma3nz_amean  \
file                                   start  end                                                            
data/Actor_24_03-01-08-02-02-02-24.wav 0 days 0 days 00:00:03.636958333                          44.371799   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file                                   start  end                                                                 
data/Actor_24_03-01-08-02-02-02-24.wav 0 days 0 days 00:00:03.636958333                                0.112749   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file                                   start  end                                                                     
data/Actor_24_03-01-08-02-02-02-24.wav 0 days 0 days 00:00:03.636958333                                   39.802952   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
file                                   start  end                                                                     
data/Actor_24_03-01-08-02-02-02-24.wav 0 days 0 days 00:00:03.636958333                                   44.422417   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile80.0  \
file                                   start  end                                                                     
data/Actor_24_03-01-08-02-02-02-24.wav 0 days 0 days 00:00:03.636958333                                   48.187508   

                                                                         F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2  \
file                                   start  end                                                                   
data/Actor_24_03-01-08-02-02-02-24.wav 0 days 0 days 00:00:03.636958333                                  8.384556   

                                                                         F0semitoneFrom27.5Hz_sma3nz_meanRisingSlope  \
file                                   start  end                                                                      
data/Actor_24_03-01-08-02-02-02-24.wav 0 days 0 days 00:00:03.636958333                                   188.801086   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevRisingSlope  \
file                                   start  end                                                                        
data/Actor_24_03-01-08-02-02-02-24.wav 0 days 0 days 00:00:03.636958333                                     277.191193   

                                                                         F0semitoneFrom27.5Hz_sma3nz_meanFallingSlope  \
file                                   start  end                                                                       
data/Actor_24_03-01-08-02-02-02-24.wav 0 days 0 days 00:00:03.636958333                                    392.898987   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevFallingSlope  \
file                                   start  end                                                                         
data/Actor_24_03-01-08-02-02-02-24.wav 0 days 0 days 00:00:03.636958333                                      663.038086   

                                                                         ...  \
file                                   start  end                        ...   
data/Actor_24_03-01-08-02-02-02-24.wav 0 days 0 days 00:00:03.636958333  ...   

                                                                         slopeUV0-500_sma3nz_amean  \
file                                   start  end                                                    
data/Actor_24_03-01-08-02-02-02-24.wav 0 days 0 days 00:00:03.6

The processing is correct, there is no error with the data processing

### Process with GeMAPSv01b

In [21]:
smile = opensmile.Smile(
    feature_set=opensmile.FeatureSet.GeMAPSv01b,
    feature_level=opensmile.FeatureLevel.Functionals,
)

from tqdm import tqdm

def generate_df(file_paths):
    dfs = []  # List to accumulate DataFrames
    
    for path in tqdm(file_paths, desc="Processing files"):
        # Assuming smile.process_file returns a DataFrame
        y = smile.process_file(path)
        dfs.append(y)  # Append the DataFrame to the list
    
    # Concatenate all DataFrames in the list
    concatenated_df = pd.concat(dfs, ignore_index=True)
    
    return concatenated_df

In [22]:
df_processed = generate_df(df['file_path'])

Processing files: 100%|████████████████████████████████████████████████████████████| 1975/1975 [03:31<00:00,  9.35it/s]


In [23]:
from sklearn.preprocessing import StandardScaler
import os

num_clusters = 9

# Set LOKY_MAX_CPU_COUNT environment variable
os.environ["LOKY_MAX_CPU_COUNT"] = "4"  # Replace "4" with the number of cores you want to use

from sklearn.cluster import KMeans

# Instantiate the StandardScaler
scaler = StandardScaler()

# Fit the scaler to your data
scaler.fit(df_processed)

# Transform the data using the scaler
df_scaled = scaler.transform(df_processed)

X_train = df_scaled
kmeans = KMeans(n_clusters=num_clusters, random_state=42)
kmeans.fit(X_train)

# Get cluster labels for each data point
cluster_labels = kmeans.labels_

from sklearn.metrics import adjusted_rand_score

# Assuming y_true contains the true labels for your data points
# cluster_labels contains the cluster assignments obtained from K-means clustering

ari = adjusted_rand_score(df['emotion'], cluster_labels)
print("Adjusted Rand Index (ARI):", ari)
print()
print('Actual Labels:',df['emotion'][:20].values)
print()
print('Unsupervised Labels', cluster_labels[:20])

Adjusted Rand Index (ARI): 0.1022215714666477

Actual Labels: ['happy' 'neutral' 'angry' 'happy' 'neutral' 'sad' 'angry' 'angry'
 'fearful' 'happy' 'boredom' 'neutral' 'sad' 'angry' 'fearful' 'happy'
 'neutral' 'sad' 'angry' 'angry']

Unsupervised Labels [6 6 4 4 6 8 4 4 6 6 6 6 8 4 6 6 6 8 6 4]


### Process with emobase

In [24]:
smile = opensmile.Smile(
    feature_set=opensmile.FeatureSet.emobase,
    feature_level=opensmile.FeatureLevel.Functionals,
)

from tqdm import tqdm

def generate_df(file_paths):
    dfs = []  # List to accumulate DataFrames
    
    for path in tqdm(file_paths, desc="Processing files"):
        # Assuming smile.process_file returns a DataFrame
        y = smile.process_file(path)
        dfs.append(y)  # Append the DataFrame to the list
    
    # Concatenate all DataFrames in the list
    concatenated_df = pd.concat(dfs, ignore_index=True)
    
    return concatenated_df

In [25]:
df_processed = generate_df(df['file_path'])

Processing files: 100%|████████████████████████████████████████████████████████████| 1975/1975 [02:04<00:00, 15.90it/s]


In [26]:
# Instantiate the StandardScaler
scaler = StandardScaler()

# Fit the scaler to your data
scaler.fit(df_processed)

# Transform the data using the scaler
df_scaled = scaler.transform(df_processed)

X_train = df_scaled
kmeans = KMeans(n_clusters=num_clusters, random_state=42)
kmeans.fit(X_train)

# Get cluster labels for each data point
cluster_labels = kmeans.labels_

from sklearn.metrics import adjusted_rand_score

# Assuming y_true contains the true labels for your data points
# cluster_labels contains the cluster assignments obtained from K-means clustering

ari = adjusted_rand_score(df['emotion'], cluster_labels)
print("Adjusted Rand Index (ARI):", ari)
print()
print('Actual Labels:',df['emotion'][:20].values)
print()
print('Unsupervised Labels', cluster_labels[:20])

Adjusted Rand Index (ARI): 0.061585928865284405

Actual Labels: ['happy' 'neutral' 'angry' 'happy' 'neutral' 'sad' 'angry' 'angry'
 'fearful' 'happy' 'boredom' 'neutral' 'sad' 'angry' 'fearful' 'happy'
 'neutral' 'sad' 'angry' 'angry']

Unsupervised Labels [4 2 4 4 2 2 4 6 4 4 2 2 2 4 4 4 2 3 4 4]


### Process with ComParE_2016

In [27]:
smile = opensmile.Smile(
    feature_set=opensmile.FeatureSet.ComParE_2016,
    feature_level=opensmile.FeatureLevel.Functionals,
)

from tqdm import tqdm

def generate_df(file_paths):
    dfs = []  # List to accumulate DataFrames
    
    for path in tqdm(file_paths, desc="Processing files"):
        # Assuming smile.process_file returns a DataFrame
        y = smile.process_file(path)
        dfs.append(y)  # Append the DataFrame to the list
    
    # Concatenate all DataFrames in the list
    concatenated_df = pd.concat(dfs, ignore_index=True)
    
    return concatenated_df

In [28]:
df_processed = generate_df(df['file_path'])

Processing files: 100%|████████████████████████████████████████████████████████████| 1975/1975 [04:50<00:00,  6.80it/s]


In [29]:
# Instantiate the StandardScaler
scaler = StandardScaler()

# Fit the scaler to your data
scaler.fit(df_processed)

# Transform the data using the scaler
df_scaled = scaler.transform(df_processed)

X_train = df_scaled
kmeans = KMeans(n_clusters=num_clusters, random_state=42)
kmeans.fit(X_train)

# Get cluster labels for each data point
cluster_labels = kmeans.labels_

from sklearn.metrics import adjusted_rand_score

# Assuming y_true contains the true labels for your data points
# cluster_labels contains the cluster assignments obtained from K-means clustering

ari = adjusted_rand_score(df['emotion'], cluster_labels)
print("Adjusted Rand Index (ARI):", ari)
print()
print('Actual Labels:',df['emotion'][:20].values)
print()
print('Unsupervised Labels', cluster_labels[:20])

Adjusted Rand Index (ARI): 0.06293955713719294

Actual Labels: ['happy' 'neutral' 'angry' 'happy' 'neutral' 'sad' 'angry' 'angry'
 'fearful' 'happy' 'boredom' 'neutral' 'sad' 'angry' 'fearful' 'happy'
 'neutral' 'sad' 'angry' 'angry']

Unsupervised Labels [3 3 4 3 3 6 3 4 3 4 6 4 6 4 3 3 3 6 4 4]


### Conclusion for Unsupervised Learning

1. K-Means clustering does not work well on the features for the standard feature sets. 0.06-0.09 accuracy even when scaling is applied.
2. This may be due to the overwhelming amount of features 62, 88, 988, 6373 which may include alot of redundant features.
3. It would be difficult to identify important features from such an overwhelming amount of features for unsupervised learning. Hence, we would explore the use State of The Art models such as wav2vec2.

### Using ANOVA to reduce the number of features

In [7]:
from sklearn.feature_selection import SelectPercentile, f_classif
from sklearn.model_selection import train_test_split
from sklearn.datasets import make_classification
import numpy as np

# Emobase feature set
smile = opensmile.Smile(
    feature_set=opensmile.FeatureSet.emobase,
    feature_level=opensmile.FeatureLevel.Functionals,
)

from tqdm import tqdm

def generate_df(file_paths):
    dfs = []  # List to accumulate DataFrames
    
    for path in tqdm(file_paths, desc="Processing files"):
        # Assuming smile.process_file returns a DataFrame
        y = smile.process_file(path)
        dfs.append(y)  # Append the DataFrame to the list
    
    # Concatenate all DataFrames in the list
    concatenated_df = pd.concat(dfs, ignore_index=True)
    
    return concatenated_df

df_processed = generate_df(df['file_path'])

Processing files: 100%|████████████████████████████████████████████████████████████| 1975/1975 [02:35<00:00, 12.67it/s]


NameError: name 'StandardScaler' is not defined

In [19]:
import os
from sklearn.preprocessing import StandardScaler

# Instantiate the StandardScaler
scaler = StandardScaler()
# Fit the scaler to your data
scaler.fit(df_processed)
# Transform the data using the scaler
df_scaled = scaler.transform(df_processed)

X = df_scaled
y = df['emotion']

# Perform ANOVA
selector = SelectPercentile(score_func=f_classif, percentile=25)
X_selected = selector.fit_transform(X, y)

# Get indices of selected features
selected_indices = selector.get_support(indices=True)
selected_features_names = df_processed.columns[selected_indices]
print(selected_features_names)

df_selected = df_processed[selected_features_names]
scaler = StandardScaler()
scaler.fit(df_selected)
# Transform the data using the scaler
df_scaled = scaler.transform(df_selected)

num_clusters = 9

# Set LOKY_MAX_CPU_COUNT environment variable
os.environ["LOKY_MAX_CPU_COUNT"] = "4"  # Replace "4" with the number of cores you want to use

from sklearn.cluster import KMeans

X_train = df_scaled
kmeans = KMeans(n_clusters=num_clusters, random_state=42)
kmeans.fit(X_train)

# Get cluster labels for each data point
cluster_labels = kmeans.labels_

from sklearn.metrics import adjusted_rand_score

# Assuming y_true contains the true labels for your data points
# cluster_labels contains the cluster assignments obtained from K-means clustering

ari = adjusted_rand_score(df['emotion'], cluster_labels)
print("Adjusted Rand Index (ARI):", ari)
print()
print('Actual Labels:',df['emotion'][:20].values)
print()
print('Unsupervised Labels', cluster_labels[:20])

C:\Users\wkwon\anaconda3\envs\dlnn\Lib\site-packages\sklearn\feature_selection\_univariate_selection.py:112: UserWarning: Features [457 479] are constant.
  warnings.warn("Features %s are constant." % constant_features_idx, UserWarning)
C:\Users\wkwon\anaconda3\envs\dlnn\Lib\site-packages\sklearn\feature_selection\_univariate_selection.py:113: RuntimeWarning: invalid value encountered in divide
  f = msb / msw
C:\Users\wkwon\anaconda3\envs\dlnn\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


Index(['pcm_intensity_sma_max', 'pcm_intensity_sma_range',
       'pcm_intensity_sma_amean', 'pcm_intensity_sma_linregc1',
       'pcm_intensity_sma_linregc2', 'pcm_intensity_sma_linregerrA',
       'pcm_intensity_sma_linregerrQ', 'pcm_intensity_sma_stddev',
       'pcm_intensity_sma_quartile1', 'pcm_intensity_sma_quartile2',
       ...
       'lspFreq_sma_de[6]_quartile3', 'lspFreq_sma_de[6]_iqr1-2',
       'lspFreq_sma_de[6]_iqr2-3', 'lspFreq_sma_de[6]_iqr1-3',
       'lspFreq_sma_de[7]_kurtosis', 'lspFreq_sma_de[7]_quartile1',
       'lspFreq_sma_de[7]_quartile3', 'lspFreq_sma_de[7]_iqr1-2',
       'lspFreq_sma_de[7]_iqr2-3', 'lspFreq_sma_de[7]_iqr1-3'],
      dtype='object', length=247)
Adjusted Rand Index (ARI): 0.08578289142855984

Actual Labels: ['happy' 'neutral' 'angry' 'happy' 'neutral' 'sad' 'angry' 'angry'
 'fearful' 'happy' 'boredom' 'neutral' 'sad' 'angry' 'fearful' 'happy'
 'neutral' 'sad' 'angry' 'angry']

Unsupervised Labels [7 5 7 7 5 1 7 6 7 7 5 5 1 7 7 7 5 1 7 7]


In [23]:
from sklearn.feature_selection import SelectPercentile, f_classif, SelectKBest
from sklearn.model_selection import train_test_split
from sklearn.datasets import make_classification
import numpy as np

# GeMAPSv01b
smile = opensmile.Smile(
    feature_set=opensmile.FeatureSet.GeMAPSv01b,
    feature_level=opensmile.FeatureLevel.Functionals,
)

from tqdm import tqdm

def generate_df(file_paths):
    dfs = []  # List to accumulate DataFrames
    
    for path in tqdm(file_paths, desc="Processing files"):
        # Assuming smile.process_file returns a DataFrame
        y = smile.process_file(path)
        dfs.append(y)  # Append the DataFrame to the list
    
    # Concatenate all DataFrames in the list
    concatenated_df = pd.concat(dfs, ignore_index=True)
    
    return concatenated_df

df_processed = generate_df(df['file_path'])

Processing files: 100%|████████████████████████████████████████████████████████████| 1975/1975 [02:48<00:00, 11.69it/s]


In [27]:
import os
from sklearn.preprocessing import StandardScaler

X = df_processed.values
y = df['emotion']

# Perform ANOVA
selector = SelectKBest(score_func=f_classif, k=30)
X_selected = selector.fit_transform(X, y)

# Get indices of selected features
selected_indices = selector.get_support(indices=True)
selected_features_names = df_processed.columns[selected_indices]
print(selected_features_names)

df_selected = df_processed[selected_features_names]
scaler = StandardScaler()
scaler.fit(df_selected)
# Transform the data using the scaler
df_scaled = scaler.transform(df_selected)

num_clusters = 9

# Set LOKY_MAX_CPU_COUNT environment variable
os.environ["LOKY_MAX_CPU_COUNT"] = "4"  # Replace "4" with the number of cores you want to use

from sklearn.cluster import KMeans

X_train = df_scaled
kmeans = KMeans(n_clusters=num_clusters, random_state=42)
kmeans.fit(X_train)

# Get cluster labels for each data point
cluster_labels = kmeans.labels_

from sklearn.metrics import adjusted_rand_score

# Assuming y_true contains the true labels for your data points
# cluster_labels contains the cluster assignments obtained from K-means clustering

ari = adjusted_rand_score(df['emotion'], cluster_labels)
print("Adjusted Rand Index (ARI):", ari)
print()
print('Actual Labels:',df['emotion'][:20].values)
print()
print('Unsupervised Labels', cluster_labels[:20])

Index(['F0semitoneFrom27.5Hz_sma3nz_amean',
       'F0semitoneFrom27.5Hz_sma3nz_percentile20.0',
       'F0semitoneFrom27.5Hz_sma3nz_percentile50.0',
       'F0semitoneFrom27.5Hz_sma3nz_percentile80.0',
       'F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2', 'loudness_sma3_amean',
       'loudness_sma3_stddevNorm', 'loudness_sma3_percentile20.0',
       'loudness_sma3_percentile50.0', 'loudness_sma3_percentile80.0',
       'loudness_sma3_pctlrange0-2', 'loudness_sma3_meanRisingSlope',
       'loudness_sma3_stddevRisingSlope', 'loudness_sma3_meanFallingSlope',
       'loudness_sma3_stddevFallingSlope', 'jitterLocal_sma3nz_amean',
       'shimmerLocaldB_sma3nz_stddevNorm', 'logRelF0-H1-A3_sma3nz_amean',
       'F1frequency_sma3nz_amean', 'F1amplitudeLogRelF0_sma3nz_amean',
       'F1amplitudeLogRelF0_sma3nz_stddevNorm', 'F2frequency_sma3nz_amean',
       'F2amplitudeLogRelF0_sma3nz_amean',
       'F2amplitudeLogRelF0_sma3nz_stddevNorm',
       'F3amplitudeLogRelF0_sma3nz_amean',
       'F3amp

C:\Users\wkwon\anaconda3\envs\dlnn\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


Adjusted Rand Index (ARI): 0.08951302793663957

Actual Labels: ['happy' 'neutral' 'angry' 'happy' 'neutral' 'sad' 'angry' 'angry'
 'fearful' 'happy' 'boredom' 'neutral' 'sad' 'angry' 'fearful' 'happy'
 'neutral' 'sad' 'angry' 'angry']

Unsupervised Labels [6 7 6 6 3 7 6 1 6 6 3 3 7 6 6 6 7 7 6 6]
